### ISSUE TITLE
Terminal Formatter Hangs in v0.7.3 when Coloring Code from STDIN

### ISSUE THREAD

**Issue by HelgeCPH** [alecthomas/chroma#378]
Hi,

Thank you for `chroma` it is quite a bit help for my work. However, I am experiencing an issue for some files that I would like to syntax-highlight with the `terminal256` formatter in v0.7.3.

For most of the files that I would like to syntax-highlight, I do not experience any issue. But for some files the formatter is hanging and I do not know why.

For example, if I call the small demo program (see `main.go` below), which reproduces the issue on `create.js` (can be received as in indicated below) with the file as an argument syntax-highlighting works nicely, whereas reading the file's contents from STDIN leaves the formatter hanging:

  * Calling `go run main.go create.js`  --> works as intended and prints the syntax-highlighted contents of `create.js` file quickly
  * `cat create.js | go run main.go` --> `main.go` hangs on line 107 of `main.go`, CPU usage increases to almost 100% of one core, and the program does not finish after running for some hours...

I am experiencing the issue both on Linux (Ubuntu 18.04 x64) and on MacOS (Mojave) with Go 1.12.6 (I cross-compiled the binaries for Linux from MacOS). But as said, the issue does not appear for all files consistently but for example for `create.js`.

The example file creating the issue:

```bash
wget https://raw.githubusercontent.com/apache/cordova-android/47daaaf14fff0da616dff364746f7519eada82f2/bin/create.js
```

I wrote the following small Go program (`main.go`) to reproduce my issue:

```go
package main

import (
	"bufio"
	"bytes"
	"fmt"
	"io/ioutil"
	"os"

	"github.com/alecthomas/chroma"
	"github.com/alecthomas/chroma/formatters"
	"github.com/alecthomas/chroma/lexers"
	"github.com/alecthomas/chroma/styles"
)

func getLexer(fname string, source string) chroma.Lexer {
	l := lexers.Match(fname)
	if l == nil {
		l = lexers.Analyse(source)
	}
	if l == nil {
		l = lexers.Fallback
	}
	l = chroma.Coalesce(l)
	return l
}

func getFormatter() chroma.Formatter {
	formatter := "terminal256"
	f := formatters.Get(formatter)
	if f == nil {
		f = formatters.Fallback
	}

	return f
}

func getStyle() *chroma.Style {
	style := "solarized-dark256"
	s := styles.Get(style)
	if s == nil {
		s = styles.Fallback
	}

	return s
}

func getContentsFromStdIn() string {
	var contentsStr string

	input := bufio.NewScanner(os.Stdin)
	for input.Scan() {
		line := input.Text()
		contentsStr += line + "\n"
	}
	return contentsStr
}

func getContentsFromFile(fileName string) string {
	var contentsStr string

	if stat, err := os.Stat(fileName); err == nil && stat.IsDir() {
		panic(err)
	}

	f, err := os.Open(fileName)
	fmt.Println(f)
	if err != nil {
		panic(err)
	}
	defer f.Close()

	contents, err := ioutil.ReadFile(f.Name())
	if err != nil {
		panic(err)
	}
	contentsStr = string(contents)
	return contentsStr
}

func main() {
	var contentsStr string
	var fileName string
	argFiles := os.Args[1:]

	if len(argFiles) == 0 {
		fileName = "/dev/stdin"
		contentsStr = getContentsFromStdIn()
	}

	if len(argFiles) == 1 {
		fileName = argFiles[0]
		contentsStr = getContentsFromFile(fileName)
	}

	lexer := getLexer(fileName, contentsStr)
	formatter := getFormatter()
	style := getStyle()

	it, err := lexer.Tokenise(nil, contentsStr)
	if err != nil {
		panic(err)
	}

	buf := new(bytes.Buffer)
	fmt.Println("I came until here quickly...")
	err = formatter.Format(buf, style, it)
	fmt.Println("There seems to be an issue in the formatter when coloring STDIN cat create.js | go run main.go it does not finish...")

	input := bufio.NewScanner(buf)

	for input.Scan() {
		line := input.Text()
		fmt.Println(line)
	}
}
```

Thank you in advance for your help.

**Comment by alecthomas** [comment]
The `chroma` command-line tool supports formatting from stdin. Have you tried it? Does it exhibit the same behaviour for you?

**Comment by alecthomas** [comment]
This is an interesting one! `chroma` (the tool) and your code, works fine with other source files, just not the one you used, for some reason.

**Comment by HelgeCPH** [comment]
Yes, the command-line tool shows the same issue. I just tried to run:

```bash
cat create.js | /Users/ropf/Documents/workspace/Go/bin/chroma --formatter="terminal256" --style="solarized-dark256"
```

It does not seem to be a formatter issue, since the following call has the issue too:

```bash
cat create.js | /Users/ropf/Documents/workspace/Go/bin/chroma
```

Both commands do not finish and use up to 100% of a processor core.

I tried a bit more to figure out if there is something in `create.js`. I cannot see anything in it directly, but it seems as if there is something around line 150, see my timings (omitted syntax-highlighted output from `chroma`, which is there in all three cases):

```
head -149 create.js | /Users/ropf/Documents/workspace/Go/bin/chroma
...
real	0m0.068s
user	0m0.058s
sys	0m0.011s

head -150 create.js | /Users/ropf/Documents/workspace/Go/bin/chroma
...
real	0m1.631s
user	0m1.588s
sys	0m0.026s

time head -151 create.js | /Users/ropf/Documents/workspace/Go/bin/chroma
...
real	0m26.243s
user	0m25.348s
sys	0m0.218s

time head -152 create.js | /Users/ropf/Documents/workspace/Go/bin/chroma
real	0m26.415s
user	0m25.763s
sys	0m0.178s

time head -153 create.js | /Users/ropf/Documents/workspace/Go/bin/chroma
real	0m50.747s
user	0m49.671s
sys	0m0.297s

time head -154 create.js | /Users/ropf/Documents/workspace/Go/bin/chroma
real	1m19.295s
user	1m17.934s
sys	0m0.440s
```

Calling it with up-to line 155 does not finish within ten minutes on my computer. However, I cannot see what that there should be an issue in the `create.js` file directly.

**Comment by alecthomas** [comment]
Interestingly, the matching itself differs when using stdin:

```
$ chroma --trace < create.js |& cut -c1-80 | head
root: pos=0, text="/*\n       Licensed to the Apache Software Foundation (ASF) u
root: pos=858, text="\n\n/*\n * create a cordova/android project\n *\n * USAGE\n
root: pos=859, text="\n/*\n * create a cordova/android project\n *\n * USAGE\n *
root: pos=860, text="/*\n * create a cordova/android project\n *\n * USAGE\n *  
root: pos=951, text="\n\nvar fso = WScript.CreateObject('Scripting.FileSystemObj
root: pos=952, text="\nvar fso = WScript.CreateObject('Scripting.FileSystemObjec
root: pos=953, text="var fso = WScript.CreateObject('Scripting.FileSystemObject'
root: pos=956, text=" fso = WScript.CreateObject('Scripting.FileSystemObject');\
root: pos=957, text="fso = WScript.CreateObject('Scripting.FileSystemObject');\n
root: pos=960, text=" = WScript.CreateObject('Scripting.FileSystemObject');\n\nf
```

vs. reading from a file:

```
$ chroma --trace create.js |& cut -c1-80 | head
root: pos=0, text="/*\n       Licensed to the Apache Software Foundation (ASF) u
slashstartsregex: pos=0, text="/*\n       Licensed to the Apache Software Founda
slashstartsregex: pos=858, text="\n\n/*\n * create a cordova/android project\n *
slashstartsregex: pos=860, text="/*\n * create a cordova/android project\n *\n *
slashstartsregex: pos=951, text="\n\nvar fso = WScript.CreateObject('Scripting.F
slashstartsregex: pos=953, text="var fso = WScript.CreateObject('Scripting.FileS
root: pos=953, text="var fso = WScript.CreateObject('Scripting.FileSystemObject'
slashstartsregex: pos=956, text=" fso = WScript.CreateObject('Scripting.FileSyst
slashstartsregex: pos=957, text="fso = WScript.CreateObject('Scripting.FileSyste
root: pos=957, text="fso = WScript.CreateObject('Scripting.FileSystemObject');\n
```

Very strange!

I verified that the content of stdin and os.Read were identical via a SHA1 hash, so there's something other than differing input going on here.

**Comment by alecthomas** [comment]
Ah okay I've tracked this down. This is a combination of two things:

1. When reading from stdin and no lexer is explicitly specified, content sniffing is attempted. In this case `Go` is (erroneously) detected because the file contains `package`.
2. The regex `"(\\\\|\\"|[^"])*"` seems to have catastrophically bad match times for this input, and there was no timeout on the regex matching.

I've added a short timeout of 250ms so that in this and similar cases, matching will just error and continue rather than take forever.

**Comment by HelgeCPH** [comment]
Thank you very much for pointing me to the issue of the wrongly chosen lexer! I fixed this on my side and should have seen it earlier especially since aIso diffed the input strings for which I could not find a difference. But I forgot to check lexer identification results.

I just pulled your latest changes and can confirm that it solves the issue for the input and the identified lexer. Thank you very much!
